parlai/agents/bart/bart.py

```diff
@@ -238,3 +239,35 @@ def _construct_token_losses(self, labels, model_output):
                 )
             )
         return token_losses
+
+    def _rank_eval_label_candidates(self, batch, batchsize):
+        """
+        Rank eval label candidates.
+
+        Overridden from TorchGeneratorAgent because BART uses
+        both EOS-BOS tokens to start decoding. During scoring of
+        candidates, we must get rid of the score for the start token.
+        """
+        cand_choices = []
+        cand_choices_scores = []
+        encoder_states = self.model.encoder(*self._encoder_input(batch))
+        for i in range(batchsize):
+            num_cands = len(batch.candidate_vecs[i])
+            enc = self.model.reorder_encoder_states(encoder_states, [i] * num_cands)
+            cands, _ = self._pad_tensor(batch.candidate_vecs[i])
+            scores, _ = self.model.decode_forced(enc, cands)
+            # ignore the score for the start token
+            scores = scores[:, 1:, :]
```

Comment:
is this the only line that is different?

I'm wondering now if we can solve all of these issues by just overriding `BartModel.output` to do this computation...

that would allow us to get rid of the duplicate `compute_loss` and `_construct_token_losses` functions as well

Reply:
the other change is that it needs reshaping...

Reply:
it looks like you just took the `reshape` out of the `cross_entropy` call? i.e. that's already in the base call

Reply:
not quite -- in the original function it has a call to `scores.view(num_cands * cands.size(1), -1)`, but this breaks with BART unless we call reshape. i think it would probably be fine to change the view call to reshape instead in TGA, but this is a slightly more expensive operation (copy instead of view). what do you think?

Reply:
since people don't seem to use this functionality too often i'll just change

Reply:
yeah given that this is rarely, if ever, used, it'll be better to just make them both `reshape`